src/main/java/com/android/volley/toolbox/JsonArrayRequest.java

```diff
@@ -37,32 +41,95 @@
      * @param errorListener Error listener, or null to ignore errors.
      */
     public JsonArrayRequest(
-            String url, Listener<JSONArray> listener, @Nullable ErrorListener errorListener) {
-        super(Method.GET, url, null, listener, errorListener);
+        String url,
+        Listener<JSONArray> listener,
+        @Nullable ErrorListener errorListener
+    ) {
+        super(
+            Method.GET,
+            url,
+            null,
+            listener,
+            errorListener
+        );
+    }
+
+    /**
+     * Creates a new request.
+     *
+     * @param method the HTTP method to use
+     * @param url URL to fetch the JSON from
+     * @param jsonRequest A {@link JSONObject} to post with the request.
+     *                    Null indicates no parameters will be posted along with request.
+     * @param listener Listener to receive the JSON response
+     * @param errorListener Error listener, or null to ignore errors.
+     */
+    public JsonArrayRequest(
+        int method,
+        String url,
+        @Nullable JSONObject jsonRequest,
+        Listener<JSONArray> listener,
+        @Nullable ErrorListener errorListener
+    ) {
+        super(
+            method,
+            url,
+            jsonRequest != null ? jsonRequest.toString() : null,
+            listener,
+            errorListener
+        );
+    }
+
+    /**
+     * Creates a new request.
+     *
+     * @param method the HTTP method to use
+     * @param url URL to fetch the JSON from
+     * @param jsonRequest A {@link JSONArray} to post with the request.
+     *                    Null indicates no parameters will be posted along with request.
+     * @param listener Listener to receive the JSON response
+     * @param errorListener Error listener, or null to ignore errors.
+     */
+    public JsonArrayRequest(
+        int method,
+        String url,
+        @Nullable JSONArray jsonRequest,
+        Listener<JSONArray> listener,
+        @Nullable ErrorListener errorListener
+    ) {
+        super(
+            method,
+            url,
+            jsonRequest != null ? jsonRequest.toString() : null,
+            listener,
+            errorListener
+        );
     }
 
     /**
      * Creates a new request.
      *
      * @param method the HTTP method to use
      * @param url URL to fetch the JSON from
-     * @param jsonRequest A {@link JSONArray} to post with the request. Null indicates no parameters
-     *     will be posted along with request.
+     * @param requestBody The content to post as the body of the request.
+     *                    Null indicates no parameters will be posted along with request.
      * @param listener Listener to receive the JSON response
      * @param errorListener Error listener, or null to ignore errors.
      */
     public JsonArrayRequest(
```

Comment:
For now, could we keep this constructor private? I'm not sure of the use case for external clients to pass in an arbitrary request body.

Reply:
I was thinking that it would allow devs to pass other kinds of data as the post body, such as plain text, form data, XML, HTML, Base64 encoded data etc. But perhaps it's best that they just implement their own subclass of Request/JsonRequest in those cases.

Dropping this constructor for now, as there would be no point to have it private.